px-map-behavior-layer-IMS.es6.js

```diff
@@ -191,13 +191,376 @@
       },
 
       /**
-       * Object to store feature id and feature pairs.
+       * Object to store the visible feature id and feature pairs.
        *
        * @type {Object}
        */
       _featureMap: {
         type: Object,
         value: {}
+      },
+
+      /**
+       * Cache of feature ids.
+       * @type {Set}
+       */
+      _featureCache: {
+        type: Set
+      },
+
+      /**
+       * An rtree cache of features.
+       * @type {Object}
+       */
+      _featureTreeCache: {
+        type: Object
+      },
+
+      /**
+       * Array of map bounds to track the cached features.
+       * @type {Array}
+       */
+      _boundsCache: {
+        type: Array
+      },
+
+      /**
+       * Defines whether features are cached to improve performance and reduce the number
+       * of request to the IMS service.
+       * @type {Boolean}
+       */
+      enableCache: {
+        type: Boolean,
+        value: true
+      },
+
+      /**
+       * Maximum size of the bounds cache. Defaults to 20.
+       * @type {Number}
+       */
+      maxBoundsCacheSize: {
+        type: Number,
+        value: 20
+      }
+    },
+
+    /**
+     * Returns whether otherBounds interacts (overlaps or touches) with bounds.
+     */
+    _boundsInteracts(bounds, otherBounds) {
+      return !(otherBounds.minX > bounds.maxX ||
+        otherBounds.minY > bounds.maxY ||
+        otherBounds.maxX < bounds.minX ||
+        otherBounds.maxY < bounds.minY);
+    },
+
+    /**
+     * Returns whether otherBounds overlaps bounds.
+     */
+    _boundsOverlaps(bounds, otherBounds) {
+      return !(otherBounds.minX >= bounds.maxX ||
+        otherBounds.minY >= bounds.maxY ||
+        otherBounds.maxX <= bounds.minX ||
+        otherBounds.maxY <= bounds.minY);
+    },
+
+    /**
+     * Returns whether otherBounds contains bounds.
+     */
+    _boundsContains(bounds, otherBounds) {
+      return !(otherBounds.minX > bounds.minX ||
+        otherBounds.minY > bounds.minY ||
+        otherBounds.maxX < bounds.maxX ||
+        otherBounds.maxY < bounds.maxY);
+    },
+
+    /**
+     * Returns an array of bounds defining the area of bounds that is not overlapped by otherBounds.
+     */
+    _splitBounds(bounds, otherBounds) {
```

Comment:
This function really needs to be broken down to make sense, perhaps move the scope of the top-level `if`, `else if` and `else` to live in their own functions as a first step.